File: Array/medium/cpp/rearrange_array_elements_by_sign.cpp
//
// Created by santosh on 12/20/25.
//

#include <bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    vector<int> rearrangeArray(vector<int> &nums)
    {
        vector<int> ans(nums.size());
        int pos = 0, neg = 1;

        for (int x : nums) {
            if (x > 0) {
                ans[pos] = x;
                pos += 2;
            } else {
                ans[neg] = x;
                neg += 2;
            }
        }
        return ans;

    }

};

int main()
{
    Solution s;
    vector<int> nums ={3,1,-2,-5,2,-4};
    vector<int> res=s.rearrangeArray(nums);
    for (int num:res)
    {
        cout<<num<<" ";
    }
    cout<<endl;
}

File: Array/medium/python/rearrange_array_elements_by_sign.py
# Rearrange array elements by sign (LeetCode 2149)
# Alternate positives and negatives starting with positive
# Created by santosh

class Solution:
    def rearrange_array(self, nums: list[int]) -> list[int]:
        """
        Two-index approach: positives at even indices, negatives at odd.
        Time: O(n), Space: O(n)
        """
        ans = [0] * len(nums)
        pos = 0
        neg = 1

        for x in nums:
            if x > 0:
                ans[pos] = x
                pos += 2
            else:
                ans[neg] = x
                neg += 2

        return ans


if __name__ == "__main__":
    s = Solution()
    nums = [3, 1, -2, -5, 2, -4]
    print(s.rearrange_array(nums))
